test/async_worker.cc

```diff
@@ -95,12 +98,65 @@ class TestWorkerNoCallback : public AsyncWorker {
   bool _succeed;
 };
 
+class EchoWorker : public AsyncWorker {
+ public:
+  EchoWorker(Function& cb, std::string& echo) : AsyncWorker(cb), echo(echo) {}
+  ~EchoWorker() {}
+
+  void Execute() override {
+    // Simulate cpu heavy task
+    std::this_thread::sleep_for(std::chrono::milliseconds(30));
+  }
+
+  void OnOK() override {
+    HandleScope scope(Env());
+    Callback().Call({Env().Null(), String::New(Env(), echo)});
+  }
+
+ private:
+  std::string echo;
+};
+
+class CancelWorker : public AsyncWorker {
+ public:
+  CancelWorker(Function& cb) : AsyncWorker(cb) {}
+  ~CancelWorker() {}
+
+  static void DoWork(const CallbackInfo& info) {
+    Function cb = info[0].As<Function>();
+    std::string echo = info[1].As<String>();
+    for (int i = 1; i < MAX_CANCEL_THREADS; i++) {
+      AsyncWorker* worker = new EchoWorker(cb, echo);
+      worker->Queue();
+      assert(worker->Env() == info.Env());
+    }
+
+    AsyncWorker* cancelWorker = new CancelWorker(cb);
+    cancelWorker->Queue();
+    cancelWorker->Cancel();
+  }
+
+  void Execute() override {
+    // Simulate cpu heavy task
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+
+  void OnOK() override {
+    NAPI_THROW_IF_FAILED_VOID(this->Env(), napi_generic_failure);
+  }
+
+  void OnError(const Error&) override {
+    NAPI_THROW_IF_FAILED_VOID(this->Env(), napi_generic_failure);
```

Comment:
It can be more intuitive if we can throw an error with an explicit message about the reason here:

```suggestion
    Napi::Error::New(env, "Unreachable").ThrowAsJavaScriptException();
```